consensus/src/simplex/actors/voter/round.rs

```diff
@@ -0,0 +1,858 @@
+                // Once we detect equivocation we clear all votes and return the
+                // leader key so the caller can surface slashable evidence.
+                let equivocator = self.record_equivocation_and_clear();
+                debug!(
+                    ?equivocator,
+                    ?new,
+                    ?previous,
+                    "notarize conflicts with certificate proposal (equivocation detected)"
+                );
+                return equivocator;
+            }
+            ProposalChange::Skipped => return None,
+        }
+        self.votes.insert_notarize(notarize);
+        None
+    }
+
+    pub fn add_verified_nullify(&mut self, nullify: Nullify<S>) {
+        self.votes.insert_nullify(nullify);
+    }
+
+    pub fn add_verified_finalize(&mut self, finalize: Finalize<S, D>) -> Option<S::PublicKey> {
+        // Finalize votes must refer to the same proposal we accepted for
+        // notarization. Replaying ProposalSlot::update here gives us the same
+        // equivocation detection guarantees as notarize handling above.
+        match self.proposal.update(&finalize.proposal, false) {
+            ProposalChange::New | ProposalChange::Unchanged => {}
+            ProposalChange::Replaced { previous, new } => {
+                let equivocator = self.record_equivocation_and_clear();
+                debug!(
+                    ?equivocator,
+                    ?new,
+                    ?previous,
+                    "finalize conflicts with certificate proposal (equivocation detected)"
+                );
+                return equivocator;
+            }
+            ProposalChange::Skipped => return None,
+        }
+        self.votes.insert_finalize(finalize);
+        None
+    }
+
+    pub fn add_verified_notarization(
+        &mut self,
+        notarization: Notarization<S, D>,
+    ) -> (bool, Option<S::PublicKey>) {
+        if self.notarization.is_some() {
+            return (false, None);
+        }
+        self.clear_deadlines();
+
+        // Certificates we recover from storage may carry a proposal that
+        // conflicts with the one we tentatively built from individual votes.
+        // `add_recovered_proposal` reruns the equivocation check and returns
+        // the leader key if the certificate proves double-signing.
+        let equivocator = self.add_recovered_proposal(notarization.proposal.clone());
+        self.notarization = Some(notarization);
+        (true, equivocator)
+    }
+
+    pub fn add_verified_nullification(&mut self, nullification: Nullification<S>) -> bool {
+        if self.nullification.is_some() {
+            return false;
+        }
+        self.clear_deadlines();
+        self.nullification = Some(nullification);
+        true
+    }
+
+    pub fn add_verified_finalization(
+        &mut self,
+        finalization: Finalization<S, D>,
+    ) -> (bool, Option<S::PublicKey>) {
```

Comment:
Remove the `bool`, it's just equal to `option.is_some()`

In fact, I might recommend `Result<(), S::PublicKey>` instead since this way, the linter will warn you that you're throwing away the equivocation information if you call `add_verified_finalization(...)` without checking the result (without, for example, at least using `let _ =` to show that you don't care about the result)

Reply:
The bool indicates the finalization was set (the equivocator may or may not be none)